onnx/defs/function.cc

```diff
@@ -104,4 +105,112 @@ FunctionBuilderRegistry& FunctionBuilderRegistry::OnnxInstance() {
   static FunctionBuilderRegistry func_builder_registry;
   return func_builder_registry;
 }
+
+std::string RenameTensorNode(
+    const std::string& node_name,
+    const std::string& internal_name) {
+  std::string new_name = "Func_" + node_name + internal_name;
+  return new_name;
+}
+
+void FunctionExpandHelper(
+    const FunctionProto& func,
+    const NodeProto& node,
+    size_t fn_counter,
+    GraphProto& g) {
+  // Create a temporary unique node prefix for tensor names
+  std::string node_name = node.has_name() ?  node.name() : func.name() + 
+    std::to_string(fn_counter);
+  int version = (int)func.since_version();
+  std::unordered_map<std::string, std::string> input_names_map;
+  std::unordered_map<std::string, std::string> output_names_map;
+  std::unordered_map<std::string, AttributeProto> attr_map;
+
+  for (int idx = 0; idx < node.input_size(); ++idx) {
+    input_names_map[func.input().Get(idx)] = node.input().Get(idx);
+  }
+  for (int idx = 0; idx < node.output_size(); ++idx) {
+    output_names_map[func.output().Get(idx)] = node.output().Get(idx);
+  }
+
+  for (auto& attr : node.attribute()) {
+    attr_map[attr.name()] = attr;
+  }
+
+  for (auto& function_node : func.node()) {
+    NodeProto* new_node = g.add_node();
+    new_node->CopyFrom(function_node);
+    new_node->clear_input();
+    new_node->clear_output();
+    new_node->clear_attribute();
+    for (auto& input : function_node.input()) {
+      if (input_names_map.count(input)) {
+        new_node->add_input(input_names_map[input]);
+      } else {
+        new_node->add_input(RenameTensorNode(node_name, input));
+      }
+    }
+    for (auto& output : function_node.output()) {
+      if (output_names_map.count(output)) {
+        new_node->add_output(output_names_map[output]);
+      } else {
+        new_node->add_output(RenameTensorNode(node_name, output));
+      }
+    }
+    for (auto& attr : function_node.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          AttributeProto* new_attr = new_node->add_attribute();
+          new_attr->CopyFrom(attr_map[attr.ref_attr_name()]);
+        }
+      } else {
+        AttributeProto* new_attr = new_node->add_attribute();
+        new_attr->CopyFrom(attr);
+      }
+    }
+  }
+}
+
+Common::Status DecomposeGraph(
```

Comment:
Is this called often? Since get_all_schemas() makes a copy of all schemas each time it's called.

Reply:
Will change it into a cache I think. store obtained schemas and get new ones when iterating thru every node.